### PR TITLE
Adding reference management for mechanism params objects

### PR DESCRIPTION
**Pull request by gamringer**
Fixing #53

### CODE REVIEW

File: pkcs11mechanism.c [hunk 1]
```diff
@@ -81,16 +92,31 @@ ZEND_END_ARG_INFO()
...
-  array_init(return_value);
-  add_assoc_long(return_value, "mechanism", o->mechanism.mechanism);
-  add_assoc_stringl(return_value, "Parameter", o->mechanism.pParameter, o->mechanism.ulParameterLen);
+    array_init(return_value);
+    add_assoc_long(return_value, "mechanism", o->mechanism.mechanism);
+    add_assoc_stringl(return_value, "Parameter", o->mechanism.pParameter, o->mechanism.ulParameterLen);
 }
 
 void pkcs11_mechanism_shutdown(pkcs11_mechanism_object *obj) {
+    if(obj->paramsObjType == "GcmParams") {
```
Comment (vjardin) [inline pkcs11mechanism.c]: You should not use a static string. Either use strcmp() or an enum/defined integer.

File: pkcs11mechanism.c [hunk 2]
```diff
@@ -50,27 +50,38 @@ PHP_METHOD(Mechanism, __construct) {
         } else if (Z_TYPE_P(mechanismArgument) == IS_OBJECT) {
             if(zend_string_equals_literal(Z_OBJ_P(mechanismArgument)->ce->name, "Pkcs11\\GcmParams")) {
                 pkcs11_gcmparams_object *mechanismParamsObj = Z_PKCS11_GCMPARAMS_P(mechanismArgument);
+                objval->paramsObj = &mechanismParamsObj;
+                objval->paramsObjType = "GcmParams";
```
Comment (vjardin) [inline pkcs11mechanism.c]: Same here.

File: pkcs11mechanism.c [hunk 3]
```diff
@@ -81,16 +92,31 @@ ZEND_END_ARG_INFO()
 
 PHP_METHOD(Mechanism, __debugInfo) {
 
-  const pkcs11_mechanism_object * const o = Z_PKCS11_MECHANISM_P(ZEND_THIS);
+    const pkcs11_mechanism_object * const o = Z_PKCS11_MECHANISM_P(ZEND_THIS);
 
-  ZEND_PARSE_PARAMETERS_NONE();
+    ZEND_PARSE_PARAMETERS_NONE();
```
Comment (vjardin) [inline pkcs11mechanism.c:96]: (minor) please avoid changing the indentation along with another purpose of your commit, even if it is a valid change.

**Review by vjardin — CHANGES_REQUESTED** [review]

File: pkcs11mechanism.c (hunk 3 above)
Comment (gamringer) [reply pkcs11mechanism.c:96]: Yeah, I even remind my own team of this at work. Cordonnier mal chaussé.

File: pkcs11mechanism.c (hunk 1 above)
Comment (gamringer) [reply pkcs11mechanism.c]: Yeah, so used to PHP...

**Comment by gamringer** [comment]
All fixed. LMK

**Comment by vjardin** [comment]
It is ok with the enum. You should squash it and git push force this serie before applying the pull request.

Moreover don't be the cordonnier and do not keep the blank lines neither the indentation changes with this pull request. 😀

**Pull request merged**